### PR TITLE
[Merged by Bors] - feat(Topology/Basic): add some `AccPt` lemmas

### PR DESCRIPTION
**Pull request by Command-Master**
Add some lemmas about `AccPt`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by Command-Master** [comment]
It's a lot nicer to rewrite `AccPt.map` using `map_nhdsWithin_le_nhdsWithin_apply` as
```
lemma AccPt.map {Y : Type*} [TopologicalSpace Y] {F : Filter Y} {x : Y}
    (h : AccPt x F) {f : Y → X} (hf1 : Continuous f) (hf2 : Function.Injective f)
    : AccPt (f x) (map f F) := by
  have : (map f (𝓝[≠] x ⊓ F)).NeBot := map_neBot (hf := h)
  simp only [Filter.map_inf hf2, map_principal] at this
  apply NeBot.mono this
  gcongr
  convert map_nhdsWithin_le_nhdsWithin_apply hf1
  ext y
  simp [hf2.ne_iff]
```
but I'm not sure how to make that work with the import hierarchy

**Review by j-loreaux — COMMENTED** [review]
I'm out of time, but I think you're missing some familiarity with the library and filters.

File: Mathlib/Topology/Basic.lean [hunk 1]
```diff
@@ -1039,6 +1049,20 @@ theorem Filter.EventuallyEq.tendsto {l : Filter α} {f : α → X} (hf : f =ᶠ[
...
+lemma map_nhds_le_nhds_apply {α : Type*} [TopologicalSpace α] {x : α}
+    {f : α → X} (hf1 : Continuous f) : map f (𝓝 x) ≤ 𝓝 (f x) := by
+  simp only [le_nhds_iff, mem_map, mem_nhds_iff]
+  intro s hs hs2
+  use f ⁻¹' s
+  simp [hs, hf1.isOpen_preimage, hs2, subset_rfl]
+
+lemma map_nhdsWithin_le_nhdsWithin_apply {α : Type*} [TopologicalSpace α]
+    {x : α} {S : Set X} {f : α → X} (hf1 : Continuous f) :
+    map f (𝓝[f ⁻¹' S] x) ≤ 𝓝[S] f x := by
```
Comment (j-loreaux) [inline Mathlib/Topology/Basic.lean:1061]: This is phrased a bit strangely (it should use `Tendsto`), but this is just:
```
  tendsto_nhdsWithin_of_tendsto_nhds_of_eventually_within _
    hf1.continuousWithinAt eventually_mem_nhdsWithin
```

File: Mathlib/Topology/Basic.lean [hunk 2]
```diff
@@ -861,6 +861,16 @@ theorem eventually_nhds_iff {p : X → Prop} :
   mem_nhds_iff.trans <| by simp only [subset_def, exists_prop, mem_setOf_eq]
 #align eventually_nhds_iff eventually_nhds_iff
 
+theorem frequently_nhds_iff {p : X → Prop} :
+    (∃ᶠ x in 𝓝 x, p x) ↔ ∀ U : Set X, IsOpen U → x ∈ U → ∃ x ∈ U, p x := by
```
[the remark this answers is not in the document]
Comment (Command-Master) [reply Mathlib/Topology/Basic.lean:865]: I thought of this as a direct analogy to `eventually_nhds_iff`. Wouldn't that also be so ppmething similar, with `(nhds_basis_opens a).eventually_iff`?

File: Mathlib/Topology/Basic.lean [hunk 3]
```diff
@@ -1039,6 +1049,20 @@ theorem Filter.EventuallyEq.tendsto {l : Filter α} {f : α → X} (hf : f =ᶠ[
     Tendsto f l (𝓝 x) :=
   tendsto_nhds_of_eventually_eq hf
 
+lemma map_nhds_le_nhds_apply {α : Type*} [TopologicalSpace α] {x : α}
+    {f : α → X} (hf1 : Continuous f) : map f (𝓝 x) ≤ 𝓝 (f x) := by
```
[the remark this answers is not in the document]
Comment (Command-Master) [reply Mathlib/Topology/Basic.lean:1053]: Hadn't thought of this like that, thanks. I'll remove it

File: Mathlib/Topology/Basic.lean [hunk 4]
```diff
@@ -862,14 +862,8 @@ theorem eventually_nhds_iff {p : X → Prop} :
...
 theorem frequently_nhds_iff {p : X → Prop} :
-    (∃ᶠ x in 𝓝 x, p x) ↔ ∀ U : Set X, IsOpen U → x ∈ U → ∃ x ∈ U, p x := by
-  simp only [frequently_iff, mem_nhds_iff, forall_exists_index, and_imp]
-  constructor
-  · intro a b c d
-    exact a b subset_rfl c d
-  · intro a _ c d e f
-    obtain ⟨v, hv, hv2⟩ := a c e f
-    exact ⟨v, d hv, hv2⟩
+    (∃ᶠ y in 𝓝 x, p y) ↔ ∀ U : Set X, IsOpen U → x ∈ U → ∃ y ∈ U, p y :=
```
Comment (j-loreaux) [inline Mathlib/Topology/Basic.lean:865]: to match the way things unfold better
```suggestion
    (∃ᶠ y in 𝓝 x, p y) ↔ ∀ U : Set X, x ∈ U → IsOpen U → ∃ y ∈ U, p y :=
```

Also, while you're here, can you please fix the statement of `eventually_nhds_iff` with the bounded / free variables the same way I had you do with this one? It's nearly impossible to parse.

File: Mathlib/Topology/ContinuousOn.lean [hunk 5]
```diff
@@ -1314,6 +1314,18 @@ theorem Continuous.piecewise {s : Set α} {f g : α → β} [∀ a, Decidable (a
...
+    (h : AccPt x F) {f : α → β} (hf1 : ContinuousAt f x) (hf2 : Function.Injective f) :
+    AccPt (f x) (map f F) := by
+  have : (Filter.map f (𝓝[≠] x ⊓ F)).NeBot := map_neBot (hf := h)
+  rw [Filter.map_inf hf2] at this
+  apply this.mono
+  apply inf_le_inf_right
+  apply tendsto_nhdsWithin_of_tendsto_nhds_of_eventually_within
+  exact hf1.continuousWithinAt
+  simp [hf2.eq_iff]
+  exact eventually_mem_nhdsWithin
```
Comment (j-loreaux) [inline Mathlib/Topology/ContinuousOn.lean:1327]: `gcongr` is a nice tactic. If you haven't seen it yet, you should definitely familiarize yourself. `simpa` is also nice.
```suggestion
  apply map_neBot (m := f) (hf := h) |>.mono
  rw [Filter.map_inf hf2]
  gcongr
  apply tendsto_nhdsWithin_of_tendsto_nhds_of_eventually_within _ hf1.continuousWithinAt
  simpa [hf2.eq_iff] using eventually_mem_nhdsWithin
```

**Review by j-loreaux — COMMENTED** [review]
getting close.

File: Mathlib/Topology/Basic.lean (hunk 2 above)
[the remark this answers is not in the document]
Comment (j-loreaux) [reply Mathlib/Topology/Basic.lean]: fair enough.

**Comment by j-loreaux** [comment]
one more thing: I realize you put this `AccPt.map` lemma here because it can't go earlier, but it seems like not a very natural place. Do you think there's another place that might be more natural (and still works)?

**Comment by Command-Master** [comment]
> one more thing: I realize you put this `AccPt.map` lemma here because it can't go earlier, but it seems like not a very natural place. Do you think there's another place that might be more natural (and still works)?

Perhaps it could go in #14834 's `DerivedSet.lean`

**Comment by j-loreaux** [comment]
Putting it in `DerivedSet.lean` sounds good. Given that, can you please:

1. remove `AccPt.map` from this PR and just leave it for that one (with the updated proof)
2. retitle this PR and update the PR description
3. ping me one final time when you're done and I'll merge it

**Comment by Command-Master** [comment]
> Putting it in `DerivedSet.lean` sounds good. Given that, can you please:
> 
>     1. remove `AccPt.map` from this PR and just leave it for that one (with the updated proof)
> 
>     2. retitle this PR and update the PR description
> 
>     3. ping me one final time when you're done and I'll merge it

Does this look good?

**Comment by j-loreaux** [comment]
bors merge

**Pull request closed without merging**